### PR TITLE
Remove debt requirement to get dividends

### PR DESCRIPTION
**Pull request by ajaya-man**
* Remove debt requirement for users to get dividends
* Use snapshot of staked amount of baln tokens
* Use on demand claim option for dividends distribution

### CODE REVIEW

File: core_contracts/dividends/dividends.py [hunk 1]
```diff
@@ -195,12 +216,17 @@ def __init__(self, db: IconScoreDatabase) -> None:
 
         self._distribution_activate = VarDB(self._DISTRIBUTION_ACTIVATE, db, value_type=bool)
```
Comment (dbrehmer) [inline core_contracts/dividends/dividends.py:217]: Need an activate method to set this to True callable only from governance.

**Review by dbrehmer — CHANGES_REQUESTED** [review]
Looking good.
Changes needed before deployment, as already known.

File: core_contracts/dividends/dividends.py [hunk 2]
```diff
@@ -489,15 +533,17 @@ def tokenFallback(self, _from: Address, _value: int, _data: bytes) -> None:
...
+                                       f"Snapshot id: {self._snapshot_id.get()}")
         self._amount_received_status.set(True)
 
     @payable
     def fallback(self):
         self._check_for_new_day()
         day = self._snapshot_id.get()
         self._daily_fees[day][str(ZERO_SCORE_ADDRESS)] += self.msg.value
-        self.DividendsReceived(self.msg.value, f"{self.msg.value} ICX received as dividends")
+        self.DividendsReceived(self.msg.value, f"{self.msg.value} ICX received as dividends. "
```
Comment (dbrehmer) [inline core_contracts/dividends/dividends.py:545]: Let's make the snapshot id the indexed datum for the DividendsReceived eventlog.

**Review by dbrehmer — APPROVED** [review]
Looks good. Only one comment on the changes:
to make the snapshot id the indexed datum for the DividendsReceived eventlog.

**Comment by dbrehmer** [comment]
There are no responses to my comments so I can't easily tell which comments have been addressed or how.

File: core_contracts/dividends/dividends.py [hunk 3]
```diff
@@ -656,22 +484,142 @@ def tokenFallback(self, _from: Address, _value: int, _data: bytes) -> None:
...
             available_tokens = loans.getAssetTokens()
-            if str(self.msg.sender) not in available_tokens.values():
-                revert(f"{TAG}: {self.msg.sender} token is not an accepted token for the dividends")
-            else:
+            if str(self.msg.sender) in available_tokens.values():
                 self._accepted_tokens.put(self.msg.sender)
+        self._check_for_new_day()
         day = self._snapshot_id.get()
         self._daily_fees[day][str(self.msg.sender)] += _value
         self.DividendsReceived(_value, f"{_value} tokens received as dividends token: {self.msg.sender}")
```
[the remark this answers is not in the document]
Comment (ajaya-man) [reply core_contracts/dividends/dividends.py]: Added Snapshot ID on the text, we can add new version of event log if we want to index the snapshot id

File: core_contracts/dividends/dividends.py [hunk 4]
```diff
@@ -656,22 +484,142 @@ def tokenFallback(self, _from: Address, _value: int, _data: bytes) -> None:
...
         self._amount_received_status.set(True)
 
     @payable
     def fallback(self):
+        self._check_for_new_day()
         day = self._snapshot_id.get()
         self._daily_fees[day][str(ZERO_SCORE_ADDRESS)] += self.msg.value
         self.DividendsReceived(self.msg.value, f"{self.msg.value} ICX received as dividends")
```
[the remark this answers is not in the document]
Comment (ajaya-man) [reply core_contracts/dividends/dividends.py]: Same as above. I will mark this as resolved.

File: core_contracts/dividends/dividends.py [hunk 5]
```diff
@@ -393,219 +390,50 @@ def setDividendsCategoryPercentage(self, _dist_list: List[DistPercentDict]) -> N
...
     @external
     def distribute(self, _activate: int = 0) -> bool:
         """
         Main method to handle the distribution of tokens to eligible BALN token holders
         :return: True if distribution has completed
         """
-        loans_score = self.create_interface_score(self._loans_score.get(), LoansInterface)
-        current_snapshot_id = loans_score.getDay()
+        self._check_for_new_day()
+        return True
```
[the remark this answers is not in the document]
Comment (ajaya-man) [reply core_contracts/dividends/dividends.py:432]: Please check the https://github.com/balancednetwork/balanced-contracts/pull/256/commits/c41fd16eecb2a2498033bafcb9aef64f58b0d3f4. 

It contains the method to transfer the daofund dividends. It needs to be called manually to trigger the transfer.

File: core_contracts/dividends/utils/consts.py [hunk 6]
```diff
@@ -17,3 +17,6 @@ class Status:
     DAOFUND_DISTRIBUTION = 5
     DISTRIBUTE_FUND_TO_HOLDERS = 6
 
+
```
[the remark this answers is not in the document]
Comment (ajaya-man) [reply core_contracts/dividends/utils/consts.py]: Removed in https://github.com/balancednetwork/balanced-contracts/pull/256/commits/4d26d1483c2556c80ed4d4bef5b6c8e3bc356827

File: core_contracts/dividends/dividends.py [hunk 7]
```diff
@@ -393,219 +390,50 @@ def setDividendsCategoryPercentage(self, _dist_list: List[DistPercentDict]) -> N
...
-        Used only to receive all fees generated by the Balanced system.
-        :param _from: Token origination address.
-        :type _from: :class:`iconservice.base.address.Address`
-        :param _value: Number of tokens sent.
-        :type _value: int
-        :param _data: Unused, ignored.
-        :type _data: bytes
+        Used only to claim all fees generated by the Balanced system.
         """
         pass
```
[the remark this answers is not in the document]
Comment (ajaya-man) [reply core_contracts/dividends/dividends.py]: Added here https://github.com/balancednetwork/balanced-contracts/pull/256/commits/92f555a5d494514fe4afd1edb6796d57470ddf19

File: core_contracts/dividends/dividends.py (hunk 1 above)
Comment (ajaya-man) [reply core_contracts/dividends/dividends.py:217]: Already present in [L230](https://github.com/balancednetwork/balanced-contracts/blob/development/core_contracts/dividends/dividends.py#L230)

File: core_contracts/dividends/dividends.py (hunk 2 above)
Comment (ajaya-man) [reply core_contracts/dividends/dividends.py:555]: okay, will add a new version of event log that would index the snapshot id.

Comment (ajaya-man) [reply core_contracts/dividends/dividends.py]: Has been addressed in https://github.com/balancednetwork/balanced-contracts/pull/256/commits/0e28d6e7992c67df6d5b0efe254ab3059b070fff

**Pull request merged**